### DISCUSSION TITLE
systemd opentabletdriver fails on system startup

### DISCUSSION THREAD

**Discussion by caseyjp11** [OpenTabletDriver/OpenTabletDriver#1392] [Q&A] [answered]
KDE Plasma Version: 5.22.5
KDE Frameworks Version: 5.85.0
Qt Version: 5.15.2
Kernel Version: 5.13.13-arch1-1 (64-bit)
Graphics Platform: X11
Processors: 6 × AMD FX(tm)-6300 Six-Core Processor
Memory: 15.6 GiB of RAM
Graphics Processor: Radeon RX550/550 Series
Huion HS610

After KDE is fully up I noticed that the tablet isn't working so:

$ systemctl --user status opentabletdriver
○ opentabletdriver.service - OpenTabletDriver Daemon
     Loaded: loaded (/usr/lib/systemd/user/opentabletdriver.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Fri 2021-09-03 23:23:02 MDT; 16min ago
    Process: 1286 ExecStart=/usr/bin/otd (code=exited, status=64)
   Main PID: 1286 (code=exited, status=64)
        CPU: 10ms

Sep 03 23:23:02 Kandalf systemd[1187]: opentabletdriver.service: Scheduled restart job, restart counter is at 5.
Sep 03 23:23:02 Kandalf systemd[1187]: Stopped OpenTabletDriver Daemon.
Sep 03 23:23:02 Kandalf systemd[1187]: opentabletdriver.service: Start request repeated too quickly.
Sep 03 23:23:02 Kandalf systemd[1187]: opentabletdriver.service: Failed with result 'exit-code'.
Sep 03 23:23:02 Kandalf systemd[1187]: Failed to start OpenTabletDriver Daemon.

-----------------------------------------------------

Once I'm sure the system is fully up and running  I issue:
systemctl --user start opentabletdriver 
and it comes right up:

$ systemctl --user status opentabletdriver
● opentabletdriver.service - OpenTabletDriver Daemon
     Loaded: loaded (/usr/lib/systemd/user/opentabletdriver.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2021-09-03 23:40:23 MDT; 8min ago
   Main PID: 7462 (otd)
      Tasks: 15 (limit: 19125)
     Memory: 76.1M
        CPU: 2.279s
     CGroup: /user.slice/user-1000.slice/user@1000.service/app.slice/opentabletdriver.service
             ├─7462 /bin/bash /usr/bin/otd
             └─7463 dotnet ./OpenTabletDriver.Daemon.dll -c ./Configurations

Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Output mode: Artist Mode
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Display area: [1920x1080@<960, 540>:0°],
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Tablet area: [254x158.75@<127, 79.375>:180°],
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Clipping: Enabled
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Ignoring reports outside area: Disabled
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Tip Binding: [OpenTabletDriver.Desktop.Binding.PenPassthroughBinding]@0%
Sep 03 23:40:27 Kandalf otd[7463]: [Evdev:Debug]        Successfully initialized virtual keyboard. (code NONE)
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Pen Bindings: Multi-Key Binding: +r -r, Multi-Key Binding: Shift+P
Sep 03 23:40:27 Kandalf otd[7463]: [Huion HS610:Info]        Express Key Bindings: Multi-Key Binding: Control+Z, Multi-Key Binding: Control+Shift+Z, Key Binding: RightBracket, Key Binding: LeftBracket, Key Binding: P, Key Binding: B, Key Binding: M, Key Binding: N, , , ,
Sep 03 23:40:27 Kandalf otd[7463]: [Settings:Info]        Driver is enabled.

----------------------------------------------------

I am currently using the git version, but tested this with the 5.3.3 release as well with the same issue.
Once up, it stays up just fine.

**Comment by InfinityGhost** [comment]
Might need to override the restart time for the unit, looks like X11 isn't ready when it tries to start.

```console
systemctl --user edit opentabletdriver.service
```

Change the `RestartSec` field to something longer, or set `StartLimitBurst` to something higher.

**Reply by caseyjp11** [reply]
I'll give that shot. As an FYI I did NOT notice this occurring while using Cinnamon. Both environments utilizing X11 with systemd handling boot/startup.

**Reply by InfinityGhost** [reply]
Sounds about right, odds are that it was just timing out as it takes longer to start KDE than Cinnamon depending on how fleshed out your setup is.

**Comment by caseyjp11** [comment]
and (sarcastic drumroll)...just like is normal with computers and code and applications...the issue quit happening...at least with the latest git version of the driver. I was never able to locate an issue either with journalctl or error logs, so I'm just going to watch how things go on boot for a bit. I'm inclined to think you are completely correct with the timing.

**Reply by InfinityGhost** [reply]
That's what I like to hear, a better experience with a development build. Glad to hear that fixed it.

**Comment by caseyjp11** [comment]
Yeah, I am as well.  I need to be on the git version for my tablet's functionality though. The release version doesn't map all the buttons on my HS610.  Of course git still doesn't have the last button or the wheel ring, but I can live without that for now. heh.

**Comment by gonX** [comment, marked as answer]
This is not something we can fix.

systemd is starting the unit before your desktop environment has populated the system user environment, resulting in OpenTabletDriver not being able to start because it doesn't know which `DISPLAY`/`WAYLAND_DISPLAY` to use.

The best fix is to make the desired unit be part of a target that is started by your WM once its startup sequence has finished. How to do this is described in great detail in OpenTabletDriver/OpenTabletDriver.Web#43.

Alternatively, simply placing `otd-daemon` in your desktop environments autostart is another option, though you lose the features that systemd's service manager provides.
